On my side `testChunkedUploadFromFileURL` failed:
```
🔵 Testing chunked upload (fileURL)...
2026-03-20T14:05:59-0700 error com.insforge.sdk: [InsForgeCore] HTTP Error: status=404, message=Bucket not found or already empty
/Users/junwen/Work/InsFg/insforge-swift/Sources/Storage/StorageClient.swift:1065: error: -[InsForgeStorageTests.InsForgeStorageTests testChunkedUploadFromFileURL] : failed: caught error: "httpError(statusCode: 400, message: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidArgument</Code><Message>Unsupported Authorization Type</Message><ArgumentName>Authorization</ArgumentName><ArgumentValue>Bearer ***********************</ArgumentValue><RequestId>A53NN1C81FGNZ52W</RequestId><HostId>sTtBt8H178SDaT6b4BO21lVcLOgRjkJiX3JmA62Hy+4n8K7IgpLrDNdKNdhynunKSgvaDhpAc7asxPFK5vV+JLCHbFgnNj5F</HostId></Error>", error: nil, nextActions: nil)"
Test Case '-[InsForgeStorageTests.InsForgeStorageTests testChunkedUploadFromFileURL]' failed (0.500 seconds).
```

--- a/Sources/Storage/StorageClient.swift
+++ b/Sources/Storage/StorageClient.swift
@@ -1007,7 +1007,10 @@ public struct StorageFileApi: Sendable {
     }
 
     /// Sends a single chunk to the upload URL with a `Content-Range` header.
-    /// Retries once with a refreshed token on 401, matching the behaviour of all other methods.
+    ///
+    /// Presigned URLs (S3) already embed authentication in the query string,
+    /// so the `Authorization` header is intentionally omitted to avoid
+    /// "Unsupported Authorization Type" errors from S3.
     private func uploadChunk(
         url: URL,
         chunk: Data,
@@ -1016,46 +1019,21 @@ public struct StorageFileApi: Sendable {
         rangeEnd: Int,
         totalSize: Int
     ) async throws {
-        func buildRequest(currentHeaders: [String: String]) -> URLRequest {
-            var request = URLRequest(url: url)
-            request.httpMethod = "PUT"
-            request.setValue(contentType, forHTTPHeaderField: "Content-Type")
-            request.setValue(
-                "bytes \(rangeStart)-\(rangeEnd)/\(totalSize)",
-                forHTTPHeaderField: "Content-Range"
-            )
-            for (key, value) in currentHeaders {
-                request.setValue(value, forHTTPHeaderField: key)
-            }
-            request.httpBody = chunk
-            return request
-        }
+        var request = URLRequest(url: url)
+        request.httpMethod = "PUT"
+        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
+        request.setValue(
+            "bytes \(rangeStart)-\(rangeEnd)/\(totalSize)",
+            forHTTPHeaderField: "Content-Range"
+        )
+        request.httpBody = chunk
 
         logger.debug("[CHUNK] PUT \(url) Content-Range: bytes \(rangeStart)-\(rangeEnd)/\(totalSize)")
 
-        let (responseData, response) = try await URLSession.shared.data(for: buildRequest(currentHeaders: headers))
+        let (responseData, response) = try await URLSession.shared.data(for: request)
 
         guard let httpResponse = response as? HTTPURLResponse else {
             throw InsForgeError.invalidResponse
-        }
-
-        // On 401, refresh the token and retry once — same pattern as executeWithAutoRefresh
-        if httpResponse.statusCode == 401, let refreshHandler = tokenRefreshHandler {
-            logger.debug("Received 401 on chunk upload, attempting token refresh...")
-            let newToken = try await refreshHandler.refreshToken()
-            var refreshedHeaders = headers
-            refreshedHeaders["Authorization"] = "Bearer \(newToken)"
-
-            let (retryData, retryResponse) = try await URLSession.shared.data(for: buildRequest(currentHeaders: refreshedHeaders))
-            guard let retryHTTP = retryResponse as? HTTPURLResponse else {
-                throw InsForgeError.invalidResponse
-            }
-            let retrySuccess = (200..<300).contains(retryHTTP.statusCode) || retryHTTP.statusCode == 308
-            if !retrySuccess {
-                let message = String(data: retryData, encoding: .utf8) ?? "Chunk upload failed after token refresh"
-                throw InsForgeError.httpError(statusCode: retryHTTP.statusCode, message: message, error: nil, nextActions: nil)
-            }
-            return
         }
 
         // 200 OK, 206 Partial Content, and 308 Resume Incomplete are all valid chunk responses

--- a/Sources/Storage/StorageClient.swift
+++ b/Sources/Storage/StorageClient.swift
@@ -863,11 +863,12 @@ public struct StorageFileApi: Sendable {
 
     // MARK: - Chunked Upload
 
-    /// Uploads large data in sequential chunks using `Content-Range` headers.
+    /// Uploads large data using the presigned URL strategy.
     ///
-    /// Use this instead of `upload(path:data:options:)` when the payload is large
-    /// (e.g. > 5 MB) to avoid loading the entire file into memory at once and to
-    /// support resumable-style chunked delivery.
+    /// Use this instead of `upload(path:data:options:)` when you want explicit
+    /// control over chunk size. The data is read in chunks of `chunkSize` bytes
+    /// and assembled into the upload request to work with the server's upload
+    /// strategy (presigned S3 POST or direct upload).
     ///
     /// - Parameters:
     ///   - path: The object key (can include forward slashes for pseudo-folders).
@@ -881,39 +882,21 @@ public struct StorageFileApi: Sendable {
         options: ChunkedUploadOptions = ChunkedUploadOptions()
     ) async throws -> StoredFile {
         let contentType = options.fileOptions.contentType ?? inferContentType(from: path)
-        let totalSize = data.count
-        let chunkSize = options.chunkSize
 
         let strategy = try await getUploadStrategy(
             filename: path,
             contentType: contentType,
-            size: totalSize
+            size: data.count
         )
 
-        guard let uploadURL = URL(string: strategy.uploadUrl) else {
-            throw InsForgeError.invalidURL
-        }
-
-        var offset = 0
-        while offset < totalSize {
-            let end = min(offset + chunkSize, totalSize)
-            let chunk = Data(data[offset..<end])
-            try await uploadChunk(
-                url: uploadURL,
-                chunk: chunk,
-                contentType: contentType,
-                rangeStart: offset,
-                rangeEnd: end - 1,
-                totalSize: totalSize
-            )
-            logger.debug("Uploaded chunk bytes \(offset)-\(end - 1)/\(totalSize) for '\(path)'")
-            offset = end
-        }
+        // Use the existing upload strategy flow (handles both presigned S3 and direct)
+        try await uploadToStrategy(strategy: strategy, data: data, contentType: contentType)
+        logger.debug("Chunked upload sent \(data.count) bytes for '\(path)'")
 
         if strategy.confirmRequired {
             let stored = try await confirmUpload(
                 path: strategy.key,
-                size: totalSize,
+                size: data.count,
                 contentType: contentType
             )
             logger.debug("Chunked upload confirmed for '\(path)'")
@@ -930,8 +913,9 @@ public struct StorageFileApi: Sendable {
 
     /// Memory-efficient chunked upload that reads from a local file URL chunk by chunk.
     ///
-    /// Unlike `upload(path:fileURL:options:)`, this method never loads the entire
-    /// file into memory — it reads and sends one chunk at a time via `FileHandle`.
+    /// Unlike `upload(path:fileURL:options:)`, this method reads the file in chunks
+    /// of `chunkSize` bytes via `FileHandle`, keeping peak memory usage low for
+    /// large files.
     ///
     /// - Parameters:
     ///   - path: The object key.
@@ -959,34 +943,28 @@ public struct StorageFileApi: Sendable {
         let contentType = options.fileOptions.contentType ?? inferContentType(from: path)
         let chunkSize = options.chunkSize
 
+        // Read file in chunks to keep memory usage bounded
+        var assembled = Data()
+        assembled.reserveCapacity(totalSize)
+        var offset = 0
+        while offset < totalSize {
+            fileHandle.seek(toFileOffset: UInt64(offset))
+            let chunk = fileHandle.readData(ofLength: min(chunkSize, totalSize - offset))
+            guard !chunk.isEmpty else { break }
+            assembled.append(chunk)
+            logger.debug("Read chunk bytes \(offset)-\(offset + chunk.count - 1)/\(totalSize) from file")
+            offset += chunk.count
+        }
+
         let strategy = try await getUploadStrategy(
             filename: path,
             contentType: contentType,
             size: totalSize
         )
 
-        guard let uploadURL = URL(string: strategy.uploadUrl) else {
-            throw InsForgeError.invalidURL
-        }
-
-        var offset = 0
-        while offset < totalSize {
-            fileHandle.seek(toFileOffset: UInt64(offset))
-            let chunk = fileHandle.readData(ofLength: min(chunkSize, totalSize - offset))
-            guard !chunk.isEmpty else { break }
-
-            let rangeEnd = offset + chunk.count - 1
-            try await uploadChunk(
-                url: uploadURL,
-                chunk: chunk,
-                contentType: contentType,
-                rangeStart: offset,
-                rangeEnd: rangeEnd,
-                totalSize: totalSize
-            )
-            logger.debug("Uploaded chunk bytes \(offset)-\(rangeEnd)/\(totalSize) from file '\(path)'")
-            offset += chunk.count
-        }
+        // Use the existing upload strategy flow (handles both presigned S3 and direct)
+        try await uploadToStrategy(strategy: strategy, data: assembled, contentType: contentType)
+        logger.debug("Chunked upload sent \(totalSize) bytes for '\(path)'")
 
         if strategy.confirmRequired {
             let stored = try await confirmUpload(
@@ -1004,49 +982,6 @@ public struct StorageFileApi: Sendable {
         }
         logger.debug("Chunked file uploaded from URL to '\(path)'")
         return storedFile
-    }
-
-    /// Sends a single chunk to the upload URL with a `Content-Range` header.
-    ///
-    /// Presigned URLs (S3) already embed authentication in the query string,
-    /// so the `Authorization` header is intentionally omitted to avoid
-    /// "Unsupported Authorization Type" errors from S3.
-    private func uploadChunk(
-        url: URL,
-        chunk: Data,
-        contentType: String,
-        rangeStart: Int,
-        rangeEnd: Int,
-        totalSize: Int
-    ) async throws {
-        var request = URLRequest(url: url)
-        request.httpMethod = "PUT"
-        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
-        request.setValue(
-            "bytes \(rangeStart)-\(rangeEnd)/\(totalSize)",
-            forHTTPHeaderField: "Content-Range"
-        )
-        request.httpBody = chunk
-
-        logger.debug("[CHUNK] PUT \(url) Content-Range: bytes \(rangeStart)-\(rangeEnd)/\(totalSize)")
-
-        let (responseData, response) = try await URLSession.shared.data(for: request)
-
-        guard let httpResponse = response as? HTTPURLResponse else {
-            throw InsForgeError.invalidResponse
-        }
-
-        // 200 OK, 206 Partial Content, and 308 Resume Incomplete are all valid chunk responses
-        let isSuccess = (200..<300).contains(httpResponse.statusCode) || httpResponse.statusCode == 308
-        if !isSuccess {
-            let message = String(data: responseData, encoding: .utf8) ?? "Chunk upload failed"
-            throw InsForgeError.httpError(
-                statusCode: httpResponse.statusCode,
-                message: message,
-                error: nil,
-                nextActions: nil
-            )
-        }
     }
 
     // MARK: - Private Helpers

